### DISCUSSION TITLE
Better where clauses type constraints

### DISCUSSION THREAD

**Discussion by seblaf1** [dotnet/csharplang#4654] [Language Ideas]
I believe the current where clauses have limitations and they could easily be much more powerful and help us better with beautiful code designs.

I suggest two things:

1. Allow customized new() constraints;
2. Allow inferred generic for type constraints in type declarations, just like method calls;

The first case is pretty straight forward. We could simply add parameterized new clauses which would allow developers to create instances of a class without specifying provider functions, as well as offering a way to ensure a generic type isn't abstract without having to provide a default constructor.

In the second case, it is better shown using an example. I have class `StateHandler<TState, TResult> where TState : State<TResult> where TResult : class`, and `State<TResult> where T Result : class`. The problem is I would like to use TResult in my StateHandler type definition without having to explicitly tell what my TResult is, rather infer it from the generic of my TState, similarly than how generic methods work. It would also allow me to not specifying constraints on TResult twice.

Currently, if you have a method like so: `MyMethod<T>(T instance)`, you can invoke that method without specifying T: `MyMethod(myObject)`. I would like to be able to specify my StateHandler type by simply specifying TState: `StateHandler<MyStateType>`, but still being able to use TResult in StateHandler. A solution would be allowing a syntax similar to this: `StateHandler<TState> where TState : State<TResult>`, where not specifying TResult gives us access to the generic type instead of imposing an additional constraint on TState.

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/769 and https://github.com/dotnet/csharplang/issues/1328
